### PR TITLE
xtext -> diag state fragment

### PR DESCRIPTION
**Pull request by MalinaStefaniaStoicanescu**
Signed-off-by: MalinaStefaniaStoicanescu <malina.stoicanescu@thalesgroup.com>

### CODE REVIEW

**Review by georgiana-ecobici — CHANGES_REQUESTED** [review]
Undo formatting on all lines!

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 1]
```diff
@@ -329,23 +334,21 @@ private void reorderCapellaFragments(Scenario scenario, EList<EObject> textEleme
...
-                interactionFragments.add(stateFragment.getStart());
-                interactionFragments.add(stateFragment.getFinish());
-              }
+
+            StateFragment stateFragment = getCorrespondingCapellaStateFragment(scenario, interactionFragments,
+                (org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) elementFromXtext);
+            if (stateFragment != null) {
+              interactionFragments.add(stateFragment.getStart());
+              interactionFragments.add(stateFragment.getFinish());
             }
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:348]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 2]
```diff
@@ -406,8 +409,9 @@ private void reorderCapellaSequenceMessages(Scenario scenario,
     });
   }
 
-  private static void editElements(Scenario scenario, EList<EObject> elements) {
+  private static void editElements(Scenario scenario, List<EObject> elements) {
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 3]
```diff
@@ -406,8 +409,9 @@ private void reorderCapellaSequenceMessages(Scenario scenario,
...
     EList<SequenceMessage> sequenceMessages = scenario.getOwnedMessages();
+    List<EObject> previousStateFragments = new ArrayList<EObject>();
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:414]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 4]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
             editElements(scenario, operand.getBlock().getBlockElements());
           }
         }
       } else if (xtextElement instanceof org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) {
-
+        previousStateFragments.add(xtextElement);
         editStateFragment(scenario,
-            (org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) xtextElement);
+            (org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) xtextElement,
+            previousStateFragments);
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:474]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 5]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
   }
-  
+
   private static void editStateFragment(Scenario scenario,
-      org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment xtextElement) {
+      org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment xtextElement,
+      List<EObject> previousEditorStateFragments) {
+
+    InstanceRole instanceRole = EmbeddedEditorInstanceHelper.getInstanceRole(xtextElement.getTimeline());
+    EList<TimeLapse> ownedTimeLapses = scenario.getOwnedTimeLapses();
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:485]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 6]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
-    String keyword = ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) xtextElement)
-        .getKeyword();
+    List<EObject> simillarStateFragments = previousEditorStateFragments.stream()
+        .filter(x -> ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getTimeline()
+            .equals(xtextElement.getTimeline())
+            && ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getName()
+                .equals(xtextElement.getName())
+            && ((org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment) x).getKeyword()
+                .equals(xtextElement.getKeyword()))
+        .collect(Collectors.toList());
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:494]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 7]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-    InstanceRole instanceRole = EmbeddedEditorInstanceHelper.getInstanceRole(timeline);
+    EObject relatedElement = getRelatedElement(scenario, instanceRole, xtextElement);
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:496]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 8]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-    EList<TimeLapse> ownedTimeLapses = scenario.getOwnedTimeLapses();
+    if (relatedElement == null)
+      return;
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:499]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 9]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-    switch (keyword) {
-    case DslConstants.STATE:
-      AbstractState abstractState = getCorrespondingCapellaAbstractState(scenario, instanceRole, name);
-      List<TimeLapse> ownedStateFragments = ownedTimeLapses.stream()
-          .filter(x -> x instanceof StateFragment && ((StateFragment) x).getRelatedAbstractState().equals(abstractState)
-              && ((StateFragment) x).getStart().getCoveredInstanceRoles().get(0).equals(instanceRole))
-          .collect(Collectors.toList());
+    List<TimeLapse> capellaStateFragments = getStateFragmentsWithGivenInstanceRoleAndRelatedElement(scenario,
+        instanceRole, xtextElement, relatedElement);
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:502]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 10]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-      if (ownedStateFragments.isEmpty()) {
-        InteractionState interactionStateStart = createInteractionState(DslConstants.START, abstractState,
-            instanceRole);
-        InteractionState interactionStateEnd = createInteractionState(DslConstants.FINISH, abstractState, instanceRole);
+    if (capellaStateFragments.isEmpty() || capellaStateFragments.size() < simillarStateFragments.size()) {
+      InteractionState interactionStateStart = createInteractionState(DslConstants.START, relatedElement, instanceRole);
+      InteractionState interactionStateEnd = createInteractionState(DslConstants.FINISH, relatedElement, instanceRole);
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:506]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 11]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-        scenario.getOwnedInteractionFragments().add(interactionStateStart);
-        scenario.getOwnedInteractionFragments().add(interactionStateEnd);
+      scenario.getOwnedInteractionFragments().add(interactionStateStart);
+      scenario.getOwnedInteractionFragments().add(interactionStateEnd);
 
-        StateFragment stateFragment = createStateFragment(interactionStateStart, interactionStateEnd, abstractState,
-            name);
+      StateFragment stateFragment = createStateFragment(interactionStateStart, interactionStateEnd, relatedElement,
+          xtextElement.getName());
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:512]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 12]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
-        ownedTimeLapses.add(stateFragment);
-      }
-      break;
-    case DslConstants.MODE:
-      break;
-    case DslConstants.FUNCTION:
-      break;
+      ownedTimeLapses.add(stateFragment);
     }
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:514]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 13]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
+      org.polarsys.capella.scenario.editor.dsl.textualScenario.StateFragment stateFragment) {
+
+    if (stateFragment.getKeyword().equals(DslConstants.FUNCTION)) {
+      return getCorrespondingCapellaAbstractFunction(scenario, instanceRole, stateFragment.getName());
+    }
+    return getCorrespondingCapellaAbstractState(scenario, instanceRole, stateFragment.getName(),
+        stateFragment.getKeyword());
+  }
+
   private static StateFragment createStateFragment(InteractionState interactionStateStart,
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:544]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 14]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
-      InteractionState interactionStateEnd, AbstractState abstractState, String name) {
+      InteractionState interactionStateEnd, EObject relatedElement, String name) {
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:546]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 15]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
     StateFragment stateFragment = InteractionFactory.eINSTANCE.createStateFragment();
-
     stateFragment.setStart(interactionStateStart);
     stateFragment.setFinish(interactionStateEnd);
-    stateFragment.setRelatedAbstractState(abstractState);
+
+    if (relatedElement instanceof AbstractFunction) {
+      stateFragment.setRelatedAbstractFunction((AbstractFunction) relatedElement);
+    } else
+      stateFragment.setRelatedAbstractState((AbstractState) relatedElement);
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:555]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 16]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
 
     return stateFragment;
   }
 
-  private static InteractionState createInteractionState(String name, AbstractState abstractState,
-      InstanceRole element) {
+  private static InteractionState createInteractionState(String name, EObject relatedElement, InstanceRole element) {
+
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:560]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 17]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
-    interactionState.setRelatedAbstractState(abstractState);
     interactionState.getCoveredInstanceRoles().add(element);
 
+    if (relatedElement instanceof AbstractFunction) {
+      interactionState.setRelatedAbstractFunction((AbstractFunction) relatedElement);
+    } else {
+      interactionState.setRelatedAbstractState((AbstractState) relatedElement);
+    }
+
     return interactionState;
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:571]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 18]
```diff
@@ -455,97 +459,146 @@ private static void editElements(Scenario scenario, EList<EObject> elements) {
...
+
+  private static AbstractState getCorrespondingCapellaAbstractState(Scenario scenario, InstanceRole element,
+      String name, String keyword) {
+    List<AbstractState> availableStates = new ArrayList<AbstractState>();
+
+    if (keyword.equals(DslConstants.STATE)) {
+      availableStates = EmbeddedEditorInstanceHelper.getStates(element);
+    } else {
+      availableStates = EmbeddedEditorInstanceHelper.getModes(element);
+    }
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:594]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 19]
```diff
@@ -616,18 +669,29 @@ private static InteractionOperand getFirstCapellaInteractionOperand(Scenario sce
...
+              && !interactionFragments.contains(((StateFragment) x).getStart()))
+          .collect(Collectors.toList());
+    } else {
+      filteredTimeLapses = scenario.getOwnedTimeLapses().stream()
+          .filter(x -> x instanceof StateFragment
+              && ((StateFragment) x).getRelatedAbstractState().getName().equals(stateFragment.getName())
+              && x.getStart().getCoveredInstanceRoles().get(0).equals(instanceRole)
+              && !interactionFragments.contains(((StateFragment) x).getStart()))
+          .collect(Collectors.toList());
+    }
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:695]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 20]
```diff
@@ -662,7 +726,8 @@ private static void cleanUpMessages(Scenario scenario, EList<EObject> messages)
     EList<SequenceMessage> sequenceMessages = scenario.getOwnedMessages();
     List<EObject> allXtextSequenceMessages = getAllXtextSequenceMessages(messages);
     List<SequenceMessage> messagesToBeDeleted = sequenceMessages.stream()
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:728]: DE AICI IN JOS, IN FISIERUL ASTA, nu mai e nicio modificare a mea

File: plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java [hunk 21]
```diff
@@ -376,8 +376,18 @@ public static String getModelContent() {
...
+
+  public static List<AbstractState> getModes(InstanceRole element) {
+    Collection<AbstractState> modesAndStates = ScenarioExt
+        .getAvailableStateModeStateFragment(element.getRepresentedInstance());
+    return modesAndStates.stream().filter(x -> x instanceof Mode).collect(Collectors.toList());
+  }
+
+  public static List<AbstractFunction> getAllocatedFunctions(InstanceRole element) {
+    return ScenarioExt.getAvailableFunctionsStateFragment(element.getRepresentedInstance()).stream()
         .collect(Collectors.toList());
```
Comment (MalinaStefaniaStoicanescu) [inline plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java:391]: EDIT

File: plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java [hunk 22]
```diff
@@ -406,8 +409,9 @@ private void reorderCapellaSequenceMessages(Scenario scenario,
     });
   }
 
-  private static void editElements(Scenario scenario, EList<EObject> elements) {
+  private static void editElements(Scenario scenario, List<EObject> elements) {
     EList<SequenceMessage> sequenceMessages = scenario.getOwnedMessages();
```
Comment (georgiana-ecobici) [inline plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java:413]: What is the reason behind changing from EList to List?

**Review by georgiana-ecobici — APPROVED** [review]

**Pull request merged**